JASP-Engine/JASP/R/MixedModelsCommon.R

```diff
@@ -97,50 +97,71 @@
 }
 
 ### common mixed-models functions
-.mmReadData      <- function(dataset, options, type = "LMM") {
-  if (!is.null(dataset)) {
-    return(dataset)
-  } else{
+.mmReadData      <- function(jaspResults, dataset, options, type = "LMM") {
+  if (is.null(dataset)) {
     if (type %in% c("LMM","BLMM")) {
-      return(readDataSetToEnd(
+      dataset <- readDataSetToEnd(
         columns.as.numeric = options$dependentVariable,
         columns = c(
           options$fixedVariables,
           options$randomVariables
         )
-      ))
+      )
     } else if (type %in% c("GLMM","BGLMM")) {
       if (options$family == "binomial_agg"){
-        return(readDataSetToEnd(
+        dataset <- readDataSetToEnd(
           columns.as.numeric = c(options$dependentVariable, options$dependentVariableAggregation),
           columns = c(
             options$fixedVariables,
             options$randomVariables
           )
-        ))
+        )
       } else  if (options$dependentVariableAggregation == "") {
-        return(readDataSetToEnd(
+        dataset <- readDataSetToEnd(
           columns.as.numeric = options$dependentVariable,
           columns = c(
             options$fixedVariables,
             options$randomVariables
           )
-        ))
+        )
       }  
     }
   }
+
+  dataset <- data.frame(dataset)
+
+  # check and use only the variables that actually used for modeling
+  used_variables <- .v(c(
+    options$dependentVariable,
+    if(type %in% c("GLMM", "BGLMM")) if(options$dependentVariableAggregation != "") options$dependentVariableAggregation,
+    unique(unlist(options$fixedEffects)),
+    if(length(options$randomVariables) != 0) options$randomVariables
+  ))
+  dataset <- dataset[,used_variables]
+  
+  # omit NAs/NaN/Infs and store the number of omitted observations
```

Comment:
There are some standard na.omit functions in R that might be useful. They are probably optimised for dropping parts of the data. @JohnnyDoorn might know more about this

Reply:
but `na.omit()` doesn't omit Inf

Reply:
Yes that's very true, I'm just thinking about performance. For my analyses I actually don't remove `Inf`, but I catch it with .hasErrors and display an error.

Reply:
Yeah, that's the good point. I changed it to `na.omit()` and error for `Inf`